core/src/state/world_state.rs

```diff
@@ -65,32 +65,59 @@ impl StateHandle {
             .write()
             .expect("Could not acquire world_state lock.")
     }
+}
 
-    /** Block asynchronously until the given sequence number is reached. */
-    pub async fn wait_for_seq(&self, sequence: u64) {
-        if self.state().logical_time >= sequence {
-            return;
-        }
-        let receiver = { self.write_state().get_listener(sequence) };
-        let _ = receiver.notified().await;
+#[derive(Debug, Clone)]
+pub struct SequenceNumberInThePast;
+
+#[derive(Debug)]
+pub struct ClosableNotifier {
+    notifier: Sender<()>,
+}
+
+#[derive(Debug)]
+pub struct ClosableNotify(Receiver<()>);
+
+impl ClosableNotify {
+    pub async fn notified(&mut self) {
+        let _ = self.0.recv().await;
+    }
+}
+
+impl ClosableNotifier {
+    fn new() -> Self {
+        let (tx, _rx) = tokio::sync::broadcast::channel(128);
+        Self { notifier: tx }
+    }
+
+    fn notify(&self) -> ClosableNotify {
+        ClosableNotify(self.notifier.subscribe())
+    }
+
+    fn notify_waiters(&mut self) {
+        let _ = self.notifier.send(());
     }
 }
 
 #[derive(Default, Debug)]
 pub struct WorldState {
     logical_time: u64,
     pub clusters: BTreeMap<ClusterName, ClusterState>,
-    listeners: HashMap<u64, Arc<Notify>>,
+    listeners: RwLock<HashMap<u64, ClosableNotifier>>,
 }
 
 impl WorldState {
-    pub fn get_listener(&mut self, sequence: u64) -> Arc<Notify> {
-        match self.listeners.entry(sequence) {
-            Entry::Occupied(entry) => entry.get().clone(),
+    pub fn get_listener(&self, sequence: u64) -> Result<ClosableNotify, SequenceNumberInThePast> {
+        if self.logical_time >= sequence {
```

Comment:
Good catch, but tather than an error in this case, should we just return a notifier which resolves immediately? Maybe `ClosableNotifier`'s `notifier` becomes an `Option`, and if it is `None`, `notified` does not await anything.